### PR TITLE
[Feature] Support replacing UDF in-place with `CREATE OR REPLACE FUNCTION`

### LINKED ISSUE #46253: Support replacing UDF in-place with `CREATE OR REPLACE FUNCTION`
- [x] https://github.com/StarRocks/starrocks/pull/46122

## Feature request

**Is your feature request related to a problem? Please describe.**
As of current, replacing a UDF is a two-step operation: DROP FUNCTION, then CREATE FUNCTION. This method involves potential downtime between the two commands for the function that is being replaced.

**Describe the solution you'd like**
Support syntax like
```sql
CREATE OR REPLACE FUNCTION function_name RETURNS return_type;
```

### PR DESCRIPTION
**Pull request by digitalpoetry**
## Why I'm doing:

As of current, replacing a UDF is a two-step operation: DROP FUNCTION, then CREATE FUNCTION. This method involves potential downtime between the two commands for the function that is being replaced.

## What I'm doing:

Fixes #46253

This change supports the following

```
CREATE OR REPLACE FUNCTION function_name RETURNS return_type;
```

The overall syntax for the [enhanced CREATE FUNCTION](https://docs.starrocks.io/docs/sql-reference/sql-statements/data-definition/CREATE_FUNCTION/) command will look like this:

```
CREATE [OR REPLACE] [GLOBAL] [AGGREGATE | TABLE] FUNCTION function_name
(arg_type [, ...])
RETURNS return_type
PROPERTIES ("key" = "value" [, ...]);
```

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by digitalpoetry** [comment]
This PR doesn't include documentation change yet. I'll push another for that once we've settled on the proposed design.

**Review by packy92 — DISMISSED** [review]

**Comment by digitalpoetry** [comment]
Thanks for the earlier review @packy92. I've since added some unit tests to increase coverage, though I've noticed the [FE Incremental Coverage Report](https://github.com/StarRocks/starrocks/pull/46122#issuecomment-2129051623) doesn't seem to update on the latest changes. Is there a manual step required to update the report?

**Review by kangkaisen — DISMISSED** [review]

**Review by packy92 — DISMISSED** [review]

**Comment by digitalpoetry** [comment]
Sorry, folks. A Sonarcloud check appeared after your approval with a duplicate code warning.

**Comment by digitalpoetry** [comment]
Looks like Sonarcloud is picking up code duplication in more than just my changes :(

**Review by kangkaisen — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/Database.java [hunk 1]
```diff
@@ -654,14 +655,14 @@ public String getCatalogName() {
         return catalogName;
     }
 
-    public synchronized void addFunction(Function function) throws UserException {
-        addFunctionImpl(function, false);
+    public synchronized void addFunction(Function function, boolean allowExists) throws UserException {
```
Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Database.java:658]: can you override this function with default value as fasle?

Comment (digitalpoetry) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Database.java:658]: sure

Comment (digitalpoetry) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Database.java:658]: although... there's only one usage of this function in the src code. Is this a matter of making the calls in the test more concise?

Comment (digitalpoetry) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Database.java:662]: done

Comment (allencelerdata) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Database.java:662]: @dirtysalt

**Review by dirtysalt — APPROVED** [review]

**Comment by digitalpoetry** [comment]
The long lead time between pushing a change and seeing check results has made this difficult. @packy92 @kangkaisen I'm hoping this is the last time I bug you for a review 🙏

**Review by kangkaisen — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**